# Approximation 

Computing in mathematics is first and foremost about _approximation_.  In this class, this word always means

> **Approximation**: Solving one hard problem by turning it into a sequence of simpler problems.  

An example of this idea is encapsulated by the statement

$$
\lim_{n\rightarrow\infty} \left(1 + \frac{1}{n}\right)^{n} = e.
$$

Now $e$ is an irrational number which to an absurd number of digits given by

$$
e = 2.71828182845904523536028747135266249775724709369995
$$

Thus, finding $e$ exactly is difficult.  However, we have a limit statement which says that the numbers $a_{n}$ where

$$
a_{n} = \left(1 + \frac{1}{n}\right)^{n}
$$

get _close to_, or _approximate_, $e$ for very, very large values of $n$.  And as you can see, you can readily use a calculator to determine $a_{n}$ for a given value of $n$.  So while knowing $e$ is difficult, using $a_{n}$ as an approximation allows us to get close.  Let's see this in action by using Python.  

In [1]:
import numpy as np

Let us represent e from above by exp in the code below, to avoid confusion.

let $$exp = (1 + 1/n)^n$$
Note the following

In [3]:
k = 1e3
print(k)

p1 = 150000000000
p2 = 1.5e11
print(p1)
print(p2)

1000.0
150000000000
150000000000.0


In [3]:
k = 1e-3
print(k)

0.001


In [5]:
def exp_approx(n):
    return (1 + 1/n)**(n)

In [11]:
exp_approx(1200)


2.717150075536425

In [5]:
print ("exp=%1.15f" %exp_approx(1e3) )

exp=2.716923932235594


In [6]:
print ("exp=%1.15f" %exp_approx(1e1) )
print ("exp=%1.15f" %exp_approx(1e2))
print ("exp=%1.15f" %exp_approx(1e3))
print ("exp=%1.15f" %exp_approx(1e4))
print ("exp=%1.15f" %exp_approx(1e5))
print ("exp=%1.15f" %exp_approx(1e6))
print ("exp=%1.15f" %exp_approx(1e7))

print ("exp=%1.15f" %exp_approx(1e10) )

exp=2.593742460100002
exp=2.704813829421528
exp=2.716923932235594
exp=2.718145926824926
exp=2.718268237192297
exp=2.718280469095753
exp=2.718281694132082
exp=2.718282053234788


**Class Activity 1**: Using the function and printing techniques above, find the 6th digit of $e$.  Said another way, don't just count on the screen.  That isn't really using a computer to its full potential.

In [7]:
# You put in code here 



So we see that, affiliated with an approximation, there is also a notion of _ error _ . Throughout the remainder of this course, by this word we mean 

> **Error**: The difference betweeen a true value or solution and an approximation to it.  

So, if we take as the "true" value of $e$ its 16-digit value

$$
e = 2.718281828459045 
$$

and we use our approximation to $e$ for $n=10^{7}$, then we can find the error between these two things.  

In [8]:
error = np.abs(2.718281828459045 - exp_approx(1e7))
print(error)

1.3432696333026684e-07


In [9]:
error = np.abs(2.718281828459045 - exp_approx(1e7))
print("The absolute error in floating point notation is: %1.15f" %error)
print("The absolute error in scientific notation is: %1.8e" %error)

The absolute error in floating point notation is: 0.000000134326963
The absolute error in scientific notation is: 1.34326963e-07


Thus we see that using $n=10^{7}$ gives us 

$$
\left|e - a_{10^{7}} \right| = 1.34326963 \times 10^{-7}
$$

or by using $n=10^{7}$, we get about 6 digits of accuracy after the decimal point.  Note, we have just computed the _absolute error_.  We can also compute the **relative error**, which in this case would be given by the quantity

$$
\frac{\left|e - a_{10^{7}} \right|}{e}. 
$$

**Class Activity 2**: Compute the _relative error_ of using $a_{10^{7}}$ as an approximation to $e$.  Do this both for floating point and scientific notations.  

In [10]:
# Your turn.  Type your answer in here. 


Unfortunately, not all approximations are created equally.  Said another way, sometimes things do not work as well as we would like.  We see that if we keep trying to use our current approximation for larger values of $n$ we get 

In [11]:
values = [1e7, 1e8, 1e9, 1e10, 1e11, 1e12, 1e13, 1e14] # Build a list of different values of n.

for num in values: # Iterate over the list of numbers value by value.
    print("e=%1.15f" %exp_approx(num)) # Find the approxmation to e using nval for n.

e=2.718281694132082
e=2.718281798347358
e=2.718282052011560
e=2.718282053234788
e=2.718282053357110
e=2.718523496037238
e=2.716110034086901
e=2.716110034087023


So, let's think about what we just did here.  First, I just introduced a `for` loop on you like whoa!  So what is this doing?  The easiest way to think about it is to realize that all it is doing is sparing us having to cut and paste more since it is exactly equivalent to having used the code

`
print "e=%1.15f" %exp_approx(1e7)
print "e=%1.15f" %exp_approx(1e8)
print "e=%1.15f" %exp_approx(1e9)
print "e=%1.15f" %exp_approx(1e10)
print "e=%1.15f" %exp_approx(1e11)
print "e=%1.15f" %exp_approx(1e12)
print "e=%1.15f" %exp_approx(1e13)
print "e=%1.15f" %exp_approx(1e14)
`

So instead of writing this out over and over again as we did above, we create what is called a _ list _ in Python, containing all the values of n we want to use.  This explains the line of code

`values = [1e7, 1e8, 1e9, 1e10, 1e11, 1e12, 1e13, 1e14]`

Then, the `for` statement iterates over the values within the list by assigning the variable `num` to be a value within `values`, executing the `print` statement.  After this is done, `num` is assigned to the next value of `values`, the `print` statement is executed, and so forth.  This is done until one reaches the end of the list `values`.

Observe that at some point, larger numbers don't necessarily give you a better approximation, but may actually be worse as seen above. So, is the limit statement wrong?  No, it is not.  What we are seeing is the limitations of trying to do math on a computer where every number has a finite amount of precision.  We will talk about this issue in far greater detail later in the course.  Suffice to say though, if we want more digits of $e$, we need a better approximation scheme.  And we get one via our very dear friend, the Taylor Series representation of $e^{x}$, which is given by

$$
e^{x} = \sum_{j=0}^{\infty} \frac{x^{j}}{j!} = 1 + x + \frac{x^{2}}{2!} + \frac{x^{3}}{3!} + \cdots
$$

Note, by $j!$ we mean

$$
j! = j(j-1)(j-2)\cdots(2)(1)(0!), ~ 0! = 1.
$$

So, in essence then, we have a new way to approximate $e$.  We get this by using the Taylor series so that 

$$
e = \sum_{j=0}^{\infty}\frac{1}{j!} = \lim_{n\rightarrow\infty}T_{n}, ~~~~ T_{n}=\sum_{j=0}^{n}\frac{1}{j!}.
$$
So what this says is that we can approxime $e$ using the truncated Taylor series $T_{n}$.  But to do this, we have to take a sum of an arbitrary number of terms in which we have to keep computing ever longer products at every term.  Clearly, a computer should come in handy here.  

The first thing we need to see is that $j!$ is what is called _ recursive _.  What is meant by that is that in order to compute $j!$, we have to compute $(j-1)!$ and then modify this result.  This in words summarizes the identity

$$
j! = j(j-1)!
$$

**Class Activity 3**: Suppose we define the terms $a_{j} = \frac{1}{j!}$.  Show, for $j\geq 0$ that 

$$
a_{j} = \left\{\begin{array}{rl}  
\frac{a_{j-1}}{j}, & j\geq 1, \\ 
1, & j=0.
\end{array}\right.
$$


Now, why fuss about this?  Well, as we are about to see, how we write the math and how we write the code can look very different.  They are always related, but frankly, they are different languages, and so in effect we are obliged to translate.  Let me show you what I mean.  Below is the code I would use to compute the partial sums above.

In [12]:
def exp_sum(tol): # A user inputs a specified tolerance given by the variable tol.
    Tj = 1 # We initialize the sum.
    aj = 1 # We initialize the term.
    j = 1 # We initialize the term number we are computing.  
    while np.abs(aj)>=tol: # Keep updating and adding terms until a term is smaller than the tolerance tol.
        aj *= 1/j # Updates the terms of the sum.
        Tj += aj # Adds the next term to the sum.
        j += 1 # Updates which term we are computing.
    return Tj

We will dissect this code in a couple of lectures.  But we can see what results we get from it.  And, yup, you guessed it, I'm going to give you more code to wrap your head around.  

In [13]:
tolvals = [1e-1, 1e-2, 1e-3, 1e-4, 1e-5, 1e-6, 1e-7] # Decide on the tolerance values we want to test.

for tolval in tolvals: # Iterate over the values within the list tolvals value by value
    print("e=%1.15f" %exp_sum(tolval)) # Find an approximation to e for the given tolerance value tol

e=2.708333333333333
e=2.716666666666666
e=2.718253968253968
e=2.718278769841270
e=2.718281525573192
e=2.718281801146385
e=2.718281826198493


Let's see what the corresponding error looks like.  

In [14]:
etrue = 2.718281828459045
tolvals = [1e-1, 1e-2, 1e-3, 1e-4, 1e-5, 1e-6, 1e-7,1e-8,1e-9,1e-10,1e-11] # Decide on the tolerance values we want to test.

for tolval in tolvals:
    print("The absolute error using a tolerance of %1.0e is: %1.5e" %(tolval,np.abs(etrue - exp_sum(tolval))) )

The absolute error using a tolerance of 1e-01 is: 9.94850e-03
The absolute error using a tolerance of 1e-02 is: 1.61516e-03
The absolute error using a tolerance of 1e-03 is: 2.78602e-05
The absolute error using a tolerance of 1e-04 is: 3.05862e-06
The absolute error using a tolerance of 1e-05 is: 3.02886e-07
The absolute error using a tolerance of 1e-06 is: 2.73127e-08
The absolute error using a tolerance of 1e-07 is: 2.26055e-09
The absolute error using a tolerance of 1e-08 is: 1.72876e-10
The absolute error using a tolerance of 1e-09 is: 1.22857e-11
The absolute error using a tolerance of 1e-10 is: 8.14904e-13
The absolute error using a tolerance of 1e-11 is: 5.01821e-14


So as we can see, we have a vastly superior method for computing $e$ by using the truncated Taylor series approach.  With that being said, we will now spend the next lecture dwelling on Taylor Series in greater detail.  